File: yeongbinim/packages/headfirst/06_command/RemoteControl.ts
import Command from './Command/Command';
import NoCommand from './Command/NoCommand';

class RemoteControl {
  private SLOT_SIZE = 7;
  private onCommands: Array<Command>;
  private offCommands: Array<Command>;

  constructor() {
    this.onCommands = Array.from(
      { length: this.SLOT_SIZE },
      () => new NoCommand(),
    );
    this.offCommands = Array.from(
      { length: this.SLOT_SIZE },
      () => new NoCommand(),
    );
  }

  public setCommand(
    slot: number,
    onCommand: Command,
    offCommand: Command,
  ): void {
    this.onCommands[slot] = onCommand;
    this.offCommands[slot] = offCommand;
  }

  public onButtonWasPushed(slot: number) {
    this.onCommands[slot].execute();
  }

  public offButtonWasPushed(slot: number) {
    this.offCommands[slot].execute();
  }

  public toString(): string {
    const buffer = [];

    buffer.push('\n---------- 리모컨 ----------\n');
    for (let i = 0; i < this.SLOT_SIZE; i++) {
      buffer.push(
        `[slot ${i}]\t${this.onCommands[i].constructor.name}\t${this.offCommands[i].constructor.name}\n`,
      );
    }
    return buffer.join('');
  }
}

export default RemoteControl;

File: jiggyjiggy/command/example/many/button/RemoteControl.java
package command.example.many.button;

import java.util.Arrays;

public class RemoteControl {
	Command[] onCommands;
	Command[] offCommands;
	Command undoCommand;
	
	public RemoteControl() {
		this.onCommands = new Command[7];
		this.offCommands = new Command[7];
		
		Command noCommand = new NoCommand();
		for (int i = 0; i < 7; i++) {
			this.onCommands[i] = noCommand;
			this.offCommands[i] = noCommand;
		}
		this.undoCommand = noCommand;
	}
	
	public void setCommand(int slot, Command onCommand, Command offCommand) {
		this.onCommands[slot] = onCommand;
		this.offCommands[slot] = offCommand;
	}
	
	public void onButtonWasPushed(int slot) {
		this.onCommands[slot].execute();
		this.undoCommand = onCommands[slot];
	}
	
	public void offButtonWasPushed(int slot) {
		this.offCommands[slot].execute();
		this.undoCommand = offCommands[slot];
	}
	
	@Override
	public String toString() {
		StringBuilder sb = new StringBuilder();
		sb.append("\n-------- Remote Control ---------- \n");
		for (int i = 0; i < onCommands.length; i++) {
			sb.append("[slot " + i + "] " + onCommands[i].getClass().getName() + "    " + offCommands[i].getClass().getName() + "\n");
		}
		return sb.toString();
	}
}
